File: 2022/Day_02/main.exs
defmodule Day_02 do
  @values %{
    "X" => 1,
    "Y" => 2,
    "Z" => 3
  }

  @winning_moves %{
    "A" => "Y",
    "B" => "Z",
    "C" => "X"
  }

  @tying_moves %{
    "A" => "X",
    "B" => "Y",
    "C" => "Z"
  }

  @losing_moves %{
    "A" => "Z",
    "B" => "X",
    "C" => "Y"
  }

  def parse_input(filename) do
    File.read!(filename)
    |> String.split("\n", trim: true)
    |> Enum.map(& String.split(&1, " ", trim: true) |> List.to_tuple)
  end

  def part1(input) do
    Enum.reduce(input, 0, fn {a,b}, acc ->
      cond do
        @winning_moves[a] == b -> 6
        @tying_moves[a] == b -> 3
        true -> 0
      end + @values[b] + acc
    end)
  end

  def part2(input) do
    Enum.reduce(input, 0, fn {a,b}, acc ->
      case b do
        "X" -> @values[@losing_moves[a]]
        "Y" -> 3 + @values[@tying_moves[a]]
        "Z" -> 6 + @values[@winning_moves[a]]
      end + acc
    end)
  end

  def main do
    filename = Enum.at(System.argv, 0, "input.txt")
    parsed_input = parse_input(filename)

    IO.inspect(part1(parsed_input), label: "Part 1")
    IO.inspect(part2(parsed_input), label: "Part 2")
  end
end

Day_02.main

File: 2022/Day_02/main.hs
import Control.Arrow ( (>>>) )

convert :: String -> Int
convert "A" = 0
convert "B" = 1
convert "C" = 2
convert "X" = 0
convert "Y" = 1
convert "Z" = 2

parseInput :: String -> IO [(Int, Int)]
parseInput filename = 
    flip fmap (readFile filename)
    $ lines >>> map ((\(a:b:_) -> (a,b)) . map convert . words)

part1 :: [(Int, Int)] -> Int
part1 = foldr (\(a,b) acc -> acc + b + 1 + 
    case () of _ | a == (b + 1) `mod` 3 -> 0
                 | a == b -> 3
                 | a == (b - 1) `mod` 3 -> 6) 0

part2 :: [(Int, Int)] -> Int
part2 = foldr (\(a,b) acc -> acc + 
    case b of 0 -> (a - 1) `mod` 3 + 1
              1 -> 3 + a + 1
              2 -> 6 + (a + 1) `mod` 3 + 1) 0

main = do
    lines <- parseInput "input.txt"
    putStrLn $ "Part 1: " ++ show (part1 lines)
    putStrLn $ "Part 2: " ++ show (part2 lines)